### DISCUSSION TITLE
Validation of custom JWT token

### DISCUSSION THREAD

**Discussion by dstibbe** [pocketbase/pocketbase#2198] [Q&A]
Let's say I'm authorized to use PocketBase through a custom OpenID implementation. I have already logged in and have granted permissions for PocketBase and I have a corresponding JWT token in my browser with the necessary claims.

However, since this is a custom OpenID implementation, installed on our own servers: how does PocketBase go about validating the signature of this JWT token? Can I provide PocketBase the necessary public key somehow?

**Comment by ganigeorgiev** [comment]
I'm not sure I follow.

PocketBase has its own JWT auth token, independently from the Access and Refresh tokens of the OAuth2 provider. Once the OAuth2 auth response is returned we no longer have use of the OAuth2 tokens.

If you want to verify the PocketBase JWT auth token in a 3rd party application, then you can send `POST /api/collections/users/auth-refresh` with `Authorization: TOKEN` header (aka. call `authRefresh()`). On invalid or expired token it will return an error. On success it will return a refreshed token with new `exp` claim and the latest user data.

**Reply by dstibbe** [reply]
Ok, clear. So, pocketbase provides its own JWT token to the user, and not that of the OpenID provider. Therefore it has no need to verify the signature based on a third party public key.

About the second part: no, we are not interested in using pocketbase as our auth provider for other services.

Other scenario: 
is it possible to have a third party auth service, not OAuth, which provides JWT tokens with claims to our pocketbase service? And all pocketbase would need to do is verify the isgnature of that JWT token based on a known public key of our third party auth service?

**Reply by ganigeorgiev** [reply]
You can [register your own custom route](https://pocketbase.io/docs/router/#register-new-route) and then authenticate from there.

To generate programmatically admin or auth record tokens you can make use of the [`tokens` package helpers](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.14.3/tokens).

**Reply by dstibbe** [reply]
Thank you 👍
